perma_web/perma/tests/test_link_count_caching.py

```diff
@@ -1,67 +1,52 @@
-from perma.models import Link, LinkUser
+def test_link_count_regular_user(link_user, link_factory):
+    """ We do some link count tallying on save """
+    link_count = link_user.link_count
+    link = link_factory(created_by=link_user, submitted_url="http://example.com")
+    link_user.refresh_from_db()
+    assert link_count + 1 == link_user.link_count
 
-from .utils import PermaTestCase
+    link.safe_delete()
+    link.save()
 
-class LinkCountCachingTestCase(PermaTestCase):
+    link_user.refresh_from_db()
+    assert link_count == link_user.link_count
 
-    @classmethod
-    def setUpTestData(cls):
-        cls.regular_user = LinkUser.objects.get(pk=4)
-        cls.org_user = LinkUser.objects.get(pk=3)
-        cls.registrar_user = LinkUser.objects.get(pk=2)
+def test_link_count_for_orgs(organization, link_user, org_link_factory):
+    """ We do some link count tallying on save. Let's make sure
+    we're adjusting the counts on the orgs """
 
-    ### TESTS ###
+    org_to_which_user_belongs = organization
+    link_user.organizations.set([org_to_which_user_belongs])
+    link_count = org_to_which_user_belongs.link_count
+    link = org_link_factory(created_by=link_user, submitted_url="http://example.com", organization=org_to_which_user_belongs)
+    link.save()
```

Comment:
I think you can simplify this by making an OrgUserFactory.

Also: I am pretty sure you can make a link associated with an org like this: `link_factory(organization=my_org, ...)`.

In which case, this would look something like:
```
def test_link_count_for_orgs(org_user, link_factory):
    """ We do some link count tallying on save. Let's make sure
    we're adjusting the counts on the orgs """

    ### TESTS ###
    org = org_user.organizations.first()
    link_count = org.link_count
    link = link_factory(created_by=link_user, submitted_url="http://example.com", organization=org)
    link.save()
    # etc.
```

A little clearer, do you think?

Reply:
I used this: https://factoryboy.readthedocs.io/en/stable/recipes.html#simple-many-to-many-relationship to make the OrgUserFactory but you still have to pass in the organization so it's not quite as simple as this. Let me know what you think and I can keep tinkering.

Reply:
The general pattern we're aiming for is to hide away the complexity inside the fixtures, so that you don't have to see it in the test.

Using that post-generation hook definitely sounds like one good approach. Another might be something in the style of the `user_with_links` fixture:

```
@pytest.fixture
def org_user(link_user, organization):
    link_user.organizations.add(organization)
    return link_user

def test_cremona(org_user):
    import pdb; pdb.set_trace()
    assert True

> /perma/perma_web/api/tests/test_link_resource.py(847)test_cremona()
-> assert True
(Pdb) org_user
<LinkUser: user0@example.com>
(Pdb) org_user.organizations.first()
<Organization: Graham, Church and White>
(Pdb) org_user.organizations.first().registrar
<Registrar: Garcia-Matthews>
```

Or, if you wanted to provide the option to pass orgs, and not always have one be randomly generated on the spot, you could do:

```
@pytest.fixture
def org_user_factory(link_user, organization):
    def f(orgs=None):
        if orgs:
            link_user.organizations.set(orgs)
        else:
            link_user.organizations.add(organization)
        return link_user
    return f

@pytest.fixture
def org_user(org_user_factory):
    return org_user_factory()
```

So then, in the test itself, you don't have to worry about constructing an org user, you can just summon one from the ether.

I remember really struggling with this pattern when Jack showed it to me, and then really enjoying it. Your mileage may vary 😄. If this doesn't strike you as a good idea, please let me know :-)

Reply:
Okay, I see what you're getting at now. Thanks for the explainer.